docs/adr/adr-008-p2p-discovery.md

```diff
@@ -0,0 +1,74 @@
+# ADR #008: Celestia-Node Full Node Discovery
+
+## Changelog
+
+- 2022.07.04 - discovery data structure
+
+## Authors
+
+@vgonkivs
+
+## Context
+
+This ADR is intended to describe p2p full node discovery in celestia node. 
+P2P discovery helps for light and full nodes to find other full nodes on the network at specified topic(`full`).
+As soon as full node is found and connection is established with it, then it will be added to a set of peers(limitedSet).
+## Decision
+
+- https://github.com/celestiaorg/celestia-node/issues/599
+
+## Detailed design
+```go
+// peersLimit is max amount of peers that will be discovered.
+peersLimit = 3
+
+// discovery combines advertise and discover services and allows to store discovered nodes.
+type discovery struct {
+	// storage where all discovered and active peers are saved.
+	set  *limitedSet
+	// libp2p.Host allows to connect to discovered peers.
+	host host.Host
+	// libp2p.Discovery allows to advertise and discover peers.
+	disc core.Discovery
+}
+
+// discoverer used to protect light nodes of being advertised as they support only peer discovery.
+type discoverer interface {
+    findPeers(ctx context.Context)
+}
+
+// limitedSet is a thread safe set of peers with given limit.
+// Inspired by libp2p peer.Set but extended with Remove method.
+type limitedSet struct {
+    lk sync.RWMutex
+    ps map[peer.ID]struct{}
+
+    limit int
+}
+```
+### Full Nodes behavior:
+1. Node starts advertising itself over DHT at `full` namespace after the system boots up in order to be found.
+2. Node starts finding other full nodes to be able to join the Full Node network.
```

Comment:
```suggestion
1. A Node starts advertising itself over DHT at `full` namespace after the system boots up in order to be found.
2. A Node starts finding other full nodes, so it can be able to join the Full Node network.
```
```suggestion
1. Node starts advertising itself over DHT at `full` namespace after the system boots up in order to be found.
2. Node starts finding other full nodes to be able to join the Full Node network.
```
```suggestion
1. Node starts advertising itself over DHT at `full` namespace after the system boots up in order to be found.
2. Node starts finding other full nodes to be able to join the Full Node network.
```